Fill missing settings with default values

gazben opened the pull request:
Currently if you deploy a new version to a server and the migrations are still running you get a `MissingSettings` exception. For long running migrations and big deployments this can be a problem.

This proposal adds a fallback option to settings defined in the `Settings` object. If you set the default in the code the settings mapper will return that default value instead of throwing an exception. For nullable values the default will be null without changing anything in the `Settings` object.

 Example:

```
    // Returns false without migrations
    public bool $mySetting = false;
    // Returns null
    public ?string $myOtherSetting;
    // Fails
    public string $myOtherSetting;
```

What do you think about this proposal?

Credits to @andrasszommer

LarsWiegers commented:
Love this and definitely a good idea

mrmmg commented:
YES, I upvote this PR!

After 3 hours of debugging this issue, I opened issue #307.

Please merge this PR into package!

freekmurze commented:
I like this, thanks! Could you add a tests and document this?

gazben commented:
@freekmurze I've added tests, and updated the docs. 

The only modification I made to the original proposal, is that by defining something nullable will not return `null` by default. There are two reasons for this:
 - 1.: I don't want to change the current behaviour
 - 2.: In php if you have a typed property in a class, normally if you try to access it, it will throw an error with `Typed property must not be accessed before initialization`, and we should stick to this.

gazben commented:
@freekmurze Will this be enough, or should I add more documentation?

freekmurze commented:
I'll take a look soon. Could you please fix the failing tests? If needed, feel free to drop older version of Laravel and / or PHP.

This might be a nice opportunity to also allow Laravel 12 👍

gazben commented:
@freekmurze I've added the missing `use` import for the tests, but the weird thing is that they are missing on the `main` branch also: https://github.com/spatie/laravel-settings/blob/main/tests/SettingsTest.php#L69C47-L69C63

The laravel 12 upgrade should be done in a separate MR I think.

gazben commented:
On it. The problem was introduced here as I see: https://github.com/spatie/laravel-settings/pull/311

And it seems the tests failed for it back then.

freekmurze commented:
Thank you!

gazben commented:
@freekmurze The `illuminate_carbon` fixes were stil included. I think they should be removed

liamseys commented:
I’m facing the same issue when attempting to load certain settings from my AppServiceProvider. I’ve tried running composer update, but the fix still doesn’t resolve the problem 🥲

https://github.com/liamseys/eagle/blob/main/app/Providers/AppServiceProvider.php

gazben commented:
You have to define defaults for the settings. As far as I see they are not defined: https://github.com/liamseys/eagle/blob/main/app/Settings/GeneralSettings.php

kaustuv90 commented:
Still doesnot work

using spatie/laravel-settings `3.4.2`

`class DocumentationSettings extends Settings
{
    public string $route_group = 'documentation';
    public bool $cache_enabled = false;
    public int $cache_period = 5;

    public static function group(): string
    {
        return 'documentation';
    }
}
`

getting all the data null also when i try to migrate the db it does not let me do that and show error that settings table does not exists

gazben commented:
In this case something is wrong because the fields are not nullable in your code. The migration issue came up to us also, it happens when you save the settings I'll make another MR to throw an exception if you want to save a missing setting.

kaustuv90 commented:
@gazben thanks. I hope this will fix the migration issue.

kaustuv90 commented:
@gazben I wonder what's wrong because I followed the docs properly. everything works well but as soon as i try to use the laravel settings in route it gives me error
